### DISCUSSION TITLE
Client Controlled Nullability: Null Propagation

### DISCUSSION THREAD

**Discussion by twof** [graphql/graphql-wg#866] [General]
[Spec changes](https://github.com/graphql/graphql-spec/pull/895)
To get a better idea of where we stand, I'll add a poll so people can vote on their favorite option. This vote is non-binding, and is only one factor in choosing which syntax we land on I'll open new threads for people to discuss each option below.

I'm going to try and separate null propagation and error handling discussion. In the original thread and during working group discussions, the conflation of the two has been causing some confusion about what options exist I think. For example, we could feasibly decide on any combination of (status quo null propagation/propagation to the nearest question mark) and (caught errors marked handled/caught errors put in a separate array/caught errors include destroyed data). 

`!` acts like Non-Nullable
[![](https://api.gh-polls.com/poll/01FT1Z9KMCFRHARMFZY9KGNJ9W/Status%20Quo)](https://api.gh-polls.com/poll/01FT1Z9KMCFRHARMFZY9KGNJ9W/Status%20Quo/vote)
Nondestructive `!`
[![](https://api.gh-polls.com/poll/01FT1Z9KMCFRHARMFZY9KGNJ9W/Nondestructive%20%60!%60)](https://api.gh-polls.com/poll/01FT1Z9KMCFRHARMFZY9KGNJ9W/Nondestructive%20%60!%60/vote)
Null propagates to nearest `?`
[![](https://api.gh-polls.com/poll/01FT1Z9KMCFRHARMFZY9KGNJ9W/Null%20propagates%20to%20nearest%20%60%3F%60)](https://api.gh-polls.com/poll/01FT1Z9KMCFRHARMFZY9KGNJ9W/Null%20propagates%20to%20nearest%20%60%3F%60/vote)

Update (March 6th, 2022): Option three was selected for development at the last working group meeting.

**Comment by twof** [comment]
## Option 1
## `!` acts like Non-Nullable
Fields marked with `!` are treated exactly like Non-Nullable schema fields. If null is found for one of those fields, then it propagates to the nearest Nullable parent.

### Example
#### Schema
```graphql
type Business {
  name: String
  isStarred: Boolean
}
```

#### Query
```graphql
{
    business {
        name
        isStarred!
    }
}
```

#### If isStarred has a `!` but returns null and business is nullable, the result will be:
```json
{
  "data": {
    "business": null
  }
}
```

#### If business has a `!`, the response will be:
```json
{
  "data": null
}
```

**Comment by twof** [comment]
## Option 2
## Nondestructive `!`
The required designator (`!`) does not change the content of `data`. It only exists to provide information to GraphQL clients. They can choose what to do when a field marked with `!` comes back null. Client behavior is likely to differ, but this flexibility can be seen as positive or a negative. This option would only change the language and validation portions of the GraphQL spec. Fields marked with `!` will have a flag attached to them to indicate that the client believes they are required.

**Reply by rivantsov** [reply]
I assume when field marked with ! returns null, there's also error object returned?

**Reply by twof** [reply]
There could be. I'm trying to separate discussions on null propagation and error handling, but I haven't got around to writing up the error handling thread yet. I'm going to try and make sure that's up a few days prior to the working group meeting.

**Reply by rivantsov** [reply]
I think it SHOULD be, since then the error is the ONLY outcome of the ! mark in the request

**Reply by twof** [reply]
> This option would only change the language and validation portions of the GraphQL spec. Fields marked with ! will have a flag attached to them to indicate that the client believes they are required.

It is not. This information can be used by clients to do things like code generation or to alter their own behavior during execution.

**Comment by twof** [comment]
## Option 3
## Null propagates to nearest `?`

In the event that a field marked with a `!` is null, rather than propagation halting at the nearest Nullable parent as in the status quo, propagation will continue to the nearest parent marked with a `?`. If a matching `?` is not supplied, null propagates all the way to `data`. For the sake of code generation, all fields between a `!` and it's matching `?` could be Non-Nullable.

### Example
#### Schema
```graphql
type Business {
  name: String
  isStarred: Boolean
}

type Owner {
  business: Business
}
```

#### Query
```graphql
{
    owner {
        business? {
            name
            isStarred!
        }
    }
}
```

#### If isStarred has a `!` but returns null and business is marked with `?`, the result will be:
```json
{
    "data": {
       "owner": {
            "business": null
        }
    }
}
```

#### Example 2
```graphql
{
    owner? {
        business {
            name
            isStarred!
        }
    }
}
```

#### Response
```json
{
    "data": {
       "owner": null
    }
}
```

#### If there is no matching `?`
```graphql
{
    owner {
        business {
            name
            isStarred!
        }
    }
}
```

```json
{
  "data": null
}
```

**Reply by twof** [reply]
> For the sake of code generation, all fields between a ! and it's matching ? could be Non-Nullable.

I'm starting the implementation for this, and I think this isn't true. One of the fields between ! and ? could come back null and we would want to propagate with the status quo rules.

**Reply by twof** [reply]
Previously when option 1 was the selected implementation, we could merge fields without matching CCN operators that were equivalent types, since all they were doing was altering the type of the field. So if a `name` field was a `String!` and a `name` field in one fragment was written `name!` and another `name` field in another fragment was written `name`, they could be merged at execution. 

Now that the CCN operators do something more than alter the type of the field I think it's impossible to merge them unless the CCN operators on them match. Open to discussion though!

**Reply by benjie** [reply]
Consider the following schema:

```graphql
type Query { a: A }
type A { b: B! }
type B { field: Int }
```

Now consider the following query:

```graphql
query {
  a {
    b? { field! }
    b { field }
  }
}
```

Were `a.b.field` to be null, the first branch would want to resolve to `{ b: null }` but the second branch to `{ b: { field: null } }`. 

**It seems clear that `?` cannot merge with a field with no operators.**

Consider:

```graphql
query {
  a? {
    b { field! }
  }
  a! {
    b { field }
  }
}
```

Were `a.b.field` to be null, the first branch desires `{ a: null }`, the second has `{a: {b: {field: null}}}`. 

**Clearly `?` cannot merge with `!`.**

---

I cannot come up with a query in which `!` cannot merge with non-`!` - it seems like either the entire document is blown up, or the section of the tree with the nearest `?` is blown up, but in either case the result is not inconsistent. Can you demonstrate a failing case for `!` vs no operators?

**Reply by twof** [reply]
Schema
```graphql
type Query { a: A }
type A { b: B }
type B { field: Int! }
```

Query
```graphql
query {
  a {
    b { field! }
  }
  a {
    b { field }
  }
}
```

Were `a.b.field` to be `null`, the first branch desires `{ data: null }`, the second has `{data: {a: {b: null}}}`.

Does that work as a counter-example, or are you saying you think it's fine to merge in that case?

**Reply by benjie** [reply]
In this case, the `a.b.field!` will throw, causing the entire result to be null. This would be the same situation as if you had this kind of thing:

```
query {
  a {
    b {
      field
      fieldThatThrows
    }
  }
}
```

where `a` and `b` are non-nullable. I.e. even if `field` was present, because `fieldThatThrows` throws the entire response is blown up, so field merging has no issues (it's either fine, or it doesn't exist because something higher up caught it). Do you see what I mean?

**Reply by twof** [reply]
I'm not sure I follow. If `a` and `b` are non-nullable, then yes, I do see how you could argue that the fields could be merged without issue because the end result would be equivalent. But in my example, the behavior of one of the two fields would have to be selected in order to merge because they would create different responses.

**Reply by benjie** [reply]
Right, but the null would only occur where a `?` had been explicitly placed, or the root. In either case, setting that location null would not invalidate the expectations of the client from a nullability perspective.

**Reply by twof** [reply]
Apologies for taking forever to respond here. I'm going to wait for [our other discussion](https://github.com/graphql/graphql-wg/discussions/1009#discussioncomment-3341916) to settle and then come back to this one since the outcome of that one will have an impact on this.

**Comment by twof** [comment]
Hello! It's been a while. The Client Controlled Nullability RFC is now winding down in favor of True Schema Nullability, so I'll be closing CCN related issues, PRs, and discussions. All further discussion should be rerouted to one of the following:
https://github.com/graphql/graphql-wg/discussions/1394
https://github.com/graphql/nullability-wg/discussions/39
https://github.com/graphql/graphql-wg/discussions/1410
https://github.com/graphql/graphql-spec/pull/1065

Additionally, we're in the process of removing the CCN experimental flag from graphql-js.

TL;DR of the new proposal(s) is that instead of allowing clients to adapt to imprecise schema definitions, our new approach is to find ways to allow schema authors to provide the precision that's been missing. There are a few different approaches to that which you can read about in the discussions above. Apollo and a couple other schema authors are in the process of putting together experiments, so if you'd like to have input or beta test the new proposal(s), reach out to us either in the Nullability Working Group, or in the #nullability-wg channel on Discord.
